File: TEST data for large repo/repo_Copy-2/Blueprism/BluePrism.Setup/Forms/SelectLanguageForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Globalization;
using System.Linq;
using System.Windows.Forms;

namespace BluePrism.Setup.Forms
{
    public partial class SelectLanguageForm : Form
    {
        public string NewLocale { get; set; }
        private bool _pseudoLocalization = false;
        private static List<LocaleListEntry> _cultureList = new List<LocaleListEntry>();
        private Font _labelFont = new Font("Segoe UI", 16, FontStyle.Regular, GraphicsUnit.Pixel);
        private Font _labelFontBold = new Font("Segoe UI", 16, FontStyle.Bold, GraphicsUnit.Pixel);
        private Helpers _helpers = new Helpers(null);
        public SelectLanguageForm(bool pseudoLocalization)
        {

            // This call is required by the Windows Form Designer.
            InitializeComponent();
            this.KeyPreview = true;
#if DEBUG
            this._pseudoLocalization = true;
#else
            this._pseudoLocalization = pseudoLocalization;
#endif
            if (_cultureList.Count == 0)
                GenerateCulturesTable();
            else
                _cultureList.FirstOrDefault(x => x.ID == 0).Text =
                    Properties.Resources.LocaleConfigForm_UsingWindowsLocaleSettings;
            _cultureList.Sort();
            PopulateLocales();
        }

        private void PopulateLocales()
        {
            ListPanel.Controls.Clear();

            var yPos = 26;
            var yImage = 0;
            var yBar = 0;
            var xImage = 126;
            var xWord = 150;
            var activeWord = "";


            if (_cultureList != null)
            {
                if (!_cultureList.Any(x => x.Active) &&
                    _cultureList.Any(x => x.Value == CultureInfo.CurrentUICulture.Name))
                    _cultureList.FirstOrDefault(x => x.Value == CultureInfo.CurrentUICulture.Name).Active = true;

                foreach (var culture in _cultureList)
                {
                    var l = new Label
                    {
                        Text = culture.Text,
                        ForeColor = Color.FromArgb(17, 126, 194),
                        AutoSize = true,
                        Name = $"L{culture.ID.ToString()}",
                        Font = _labelFont
                    };
                    l.Click += L_Click;
                    ListPanel.Controls.Add(l);
                    l.Location = new Point(xWord, yPos);
                    if (culture.Active)
                    {
                        activeWord = l.Text;
                        yImage = yPos + 4;
                        yBar = yPos + 24;
                    }

                    yPos += 32;
                }
            }

            var i = new PictureBox
            {
                Image = Properties.Resources.triangle_fwd_2x,
                Size = new Size(16, 16),
                Name = $"ActiveEntry",
                Visible = (yImage > 0)
            };
            ListPanel.Controls.Add(i);
            i.Location = new Point(xImage, yImage);

            var b = new Panel()
            {
                BackColor = Color.FromArgb(17, 126, 194),
                Height = 4,
                Width = _helpers.MeasureString(_labelFontBold, activeWord),
                Name = "UnderlineBar",
                Visible = (yBar > 0)
            };
            ListPanel.Controls.Add(b);
            b.Location = new Point(xWord + 5, yBar);

        }

        private void L_Click(object sender, EventArgs e)
        {
            var label = (Label)sender;
            var selectedLocale = label.Name.Substring(1);
            var oldCulture = _cultureList.FirstOrDefault(x => x.Active);

            if (oldCulture != null)
            {
                _cultureList.Remove(oldCulture);
                oldCulture.Active = false;
                _cultureList.Add(oldCulture);
            }

            var newCulture = _cultureList.FirstOrDefault(x => x.ID.ToString() == selectedLocale);

            if (newCulture != null)
            {
                ResetLabelFonts();
                _cultureList.Remove(newCulture);
                newCulture.Active = true;
                _cultureList.Add(newCulture);

                if (newCulture.ID.ToString() != null)
                {
                    var activeEntry = (PictureBox)ListPanel.Controls.Find("ActiveEntry", true).First();
                    activeEntry.Visible = true;
                    activeEntry.Location = new Point(126, label.Location.Y + 4);

                    var underlineBar = (Panel)ListPanel.Controls.Find("UnderlineBar", true).First();
                    underlineBar.Location = new Point(label.Location.X + 5, label.Location.Y + 24);
                    underlineBar.Width = _helpers.MeasureString(_labelFontBold, label.Text);
                    underlineBar.Visible = true;

                    label.Font = _labelFontBold;
                    label.AutoSize = true;
                }

                NewLocale = newCulture.Value;
            }

            _cultureList.Sort();
        }

        private void ResetLabelFonts()
        {
            foreach (Control c in ListPanel.Controls)
            {
                if (c is Label label)
                    label.Font = _labelFont;
            }
        }


        private class LocaleListEntry : IComparable
        {
            public int ID;
            public string Text;
            public string Value;
            public bool Active;

            public int CompareTo(object obj)
            {
                return ID.CompareTo(((LocaleListEntry)obj).ID);
            }
        }

        private class LocaleListComparer : IComparer<LocaleListEntry>
        {
            public int Compare(LocaleListEntry x, LocaleListEntry y)
            {
                return x.ID.CompareTo(y.ID);
            }
        }
        /// <summary>
        ///     ''' Generates list of supported locales
        ///     ''' </summary>
        private void GenerateCulturesTable()
        {
            _cultureList.Clear();
            var id = 0;
            var installedCulture = CultureInfo.CurrentUICulture;
            _cultureList.Add(new LocaleListEntry { ID = id, Text = Properties.Resources.LocaleConfigForm_UsingWindowsLocaleSettings, Value = installedCulture.Name, Active = installedCulture.Name == NewLocale });

            foreach (string locale in Internationalisation.Locales.SupportedLocales)
            {
                id++;
                try
                {
                    var addCulture = new CultureInfo(locale);
                    _cultureList.Add(new LocaleListEntry { ID = id, Text = addCulture.TextInfo.ToTitleCase(addCulture.NativeName), Value = addCulture.Name, Active = addCulture.Name == NewLocale });
                }
                catch
                {
                    // do nothing, just don't add to the drop down
                }
            }

            if (_pseudoLocalization)
            {
                try
                {
                    var addCulture = new CultureInfo("gsw-LI");
                    _cultureList.Add(new LocaleListEntry { ID = _cultureList.Count, Text = "i18n TEST LANGUAGE", Value = addCulture.Name, Active = addCulture.Name == NewLocale });
                }
                catch
                {
                    // do nothing, just don't add to the drop down
                }
            }
        }

        private void BtnOK_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.OK;
            Close();
        }

        private void BtnCancel_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
            Close();
        }

        #region "Drag And Drop Window"
        private Point _MouseDownLocation;

        private void BorderPanel_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left) _MouseDownLocation = e.Location;
        }

        private void BorderPanel_MouseMove(object sender, MouseEventArgs e)
        {
            if (e.Button != MouseButtons.Left) return;

            var form = (Form)this;
            form.Left += e.Location.X - _MouseDownLocation.X;
            form.Top += e.Location.Y - _MouseDownLocation.Y;
        }

        #endregion

        private void BorderPanel_MouseUp(object sender, MouseEventArgs e)
        {
            //following on from a drag and drop of the screen the ListPanel scrolbar become unselectable (mouse scrolling works)
            //but tweaking the vertical scroll value this causes the control to become responsive again, invalidating or refreshing does not work.
            if (e.Button != MouseButtons.Left) return;

            ListPanel.VerticalScroll.Value =
                ListPanel.VerticalScroll.Value + 1 < ListPanel.VerticalScroll.Maximum
                    ? ListPanel.VerticalScroll.Value + 1
                    : ListPanel.VerticalScroll.Maximum;
            ListPanel.VerticalScroll.Value =
                ListPanel.VerticalScroll.Value - 1 > ListPanel.VerticalScroll.Minimum
                    ? ListPanel.VerticalScroll.Value - 1
                    : ListPanel.VerticalScroll.Minimum;
        }

        private void SelectLanguageForm_KeyDown(object sender, KeyEventArgs e)
        {
            switch (e.KeyCode)
            {
                case Keys.Down:
                    {
                        SelectNextLocale(1);
                        break;
                    }

                case Keys.Up:
                    {
                        SelectNextLocale(-1);
                        break;
                    }
            }
        }

        private void SelectNextLocale(int dir)
        {
            var oldCulture = _cultureList.FirstOrDefault(x => x.Active);
            var index = oldCulture.ID + dir;
            if (ListPanel.Controls.Find($"L{index}", true).FirstOrDefault() is Label label)
            {
                ListPanel.ScrollControlIntoView(label);
                L_Click(label, EventArgs.Empty);
            }
        }
    }
}

File: TEST data for large repo/Repo Copy-3/Blueprism/BluePrism.Automate/AutomateConfig/UI/SelectLanguageForm.vb
Imports System.Globalization
Imports System.Linq
Imports BluePrism.AutomateAppCore
Imports BluePrism.Core.Utility

Public Class SelectLanguageForm
    Public NewLocale As String = Nothing
    Private ReadOnly mPseudoLocalization As Boolean = False
    Private Shared mCultureList As List(Of LocaleListEntry) = New List(Of LocaleListEntry)()
    Private ReadOnly mLabelFont As Font = New Font("Segoe UI", 16, FontStyle.Regular, GraphicsUnit.Pixel)
    Private ReadOnly mLabelFontBold As Font = New Font("Segoe UI", 16, FontStyle.Bold, GraphicsUnit.Pixel)

    Public Sub New(ByVal pseudoLocalization As Boolean)
        InitializeComponent()
        Me.KeyPreview = True
        mPseudoLocalization = pseudoLocalization
#If DEBUG Then
        mPseudoLocalization = True
#End If
        If mCultureList.Count = 0 Then
            GenerateCulturesTable()
        Else
            mCultureList.FirstOrDefault(Function(x) x.ID = 0).Text = My.Resources.LocaleConfigForm_UsingWindowsLocaleSettings
        End If
        mCultureList.Sort()
        PopulateLocales()
    End Sub

    Private Sub PopulateLocales()
        ListPanel.Controls.Clear()
        Dim yPos = 26
        Dim yImage = 0
        Dim yBar = 0
        Dim xImage = 126
        Dim xWord = 150
        Dim activeWord = ""

        If mCultureList.Where(Function(x) x.Active).Count = 0 Then
            Dim currentCulture = CultureInfo.CurrentUICulture.Name
            If CultureInfo.CurrentUICulture.Parent.Name = "fr" Then currentCulture = "fr-FR"
            If CultureHelper.IsLatinAmericanSpanish() Then currentCulture = CultureHelper.LatinAmericanSpanish
            If mCultureList.FirstOrDefault(Function(x) x.Value = currentCulture) IsNot Nothing Then
                mCultureList.FirstOrDefault(Function(x) x.Value = currentCulture).Active = True
            End If
        End If

        For Each culture In mCultureList
            Dim l = New Label With {
                    .Text = culture.Text,
                    .ForeColor = Color.FromArgb(17, 126, 194),
                    .Size = New Size(clsUtility.MeasureString(mLabelFontBold, culture.Text), 24),
                    .Name = $"L{culture.ID.ToString()}",
                    .Font = mLabelFont
                }
            AddHandler l.Click, AddressOf L_Click
            ListPanel.Controls.Add(l)
            l.Location = New Point(xWord, yPos)

            If culture.Active Then
                activeWord = l.Text
                yImage = yPos + 4
                yBar = yPos + 24
                l.Font = mLabelFontBold
            End If

            yPos += 32
        Next

        Dim i = New PictureBox With {
                .Image = My.Resources.triangle_fwd_2x,
                .Size = New Size(16, 16),
                .Name = $"ActiveEntry",
                .Visible = (yImage > 0)
            }
        ListPanel.Controls.Add(i)
        i.Location = New Point(xImage, yImage)
        Dim b = New Panel() With {
                .BackColor = Color.FromArgb(17, 126, 194),
                .Height = 4,
                .Width = MeasureString(mLabelFontBold, activeWord),
                .Name = "UnderlineBar",
                .Visible = (yBar > 0)
            }
        ListPanel.Controls.Add(b)
        b.Location = New Point(xWord + 5, yBar)
    End Sub

    Private Function MeasureString(lFont As Font, lText As String) As Integer
        Using image = New Bitmap(1, 1)
            Using g = Graphics.FromImage(image)
                Return CInt(Math.Ceiling(g.MeasureString(lText, lFont).Width))
            End Using
        End Using
    End Function

    Private Sub L_Click(sender As Object, e As EventArgs)

        Dim label = CType(sender, Label)
        Dim selectedLocale = label.Name.Substring(1)
        Dim oldCulture = mCultureList.FirstOrDefault(Function(x) x.Active)

        If oldCulture IsNot Nothing Then
            mCultureList.Remove(oldCulture)
            oldCulture.Active = False
            mCultureList.Add(oldCulture)
        End If

        Dim newCulture = mCultureList.FirstOrDefault(Function(x) x.ID.ToString() = selectedLocale)

        If newCulture IsNot Nothing Then
            ResetLabelFonts()
            mCultureList.Remove(newCulture)
            newCulture.Active = True
            mCultureList.Add(newCulture)

            If newCulture.ID.ToString() IsNot Nothing Then
                Dim activeEntry = CType(ListPanel.Controls.Find("ActiveEntry", True).First(), PictureBox)
                activeEntry.Visible = True
                activeEntry.Location = New Point(126, label.Location.Y + 4)

                Dim underlineBar = CType(ListPanel.Controls.Find("UnderlineBar", True).First(), Panel)
                underlineBar.Location = New Point(label.Location.X + 5, label.Location.Y + 24)
                underlineBar.Width = MeasureString(mLabelFontBold, label.Text)
                underlineBar.Visible = True

                label.Font = mLabelFontBold
            End If

            NewLocale = newCulture.Value
        End If

        mCultureList.Sort()
    End Sub

    Private Sub ResetLabelFonts()
        For Each c As Control In ListPanel.Controls
            Dim label = TryCast(c, Label)
            If label IsNot Nothing Then label.Font = mLabelFont
        Next
    End Sub

    Private Class LocaleListEntry
        Implements IComparable

        Public ID As Integer
        Public Text As String
        Public Value As String
        Public Active As Boolean


        Private Function IComparable_CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
            Return ID.CompareTo(CType(obj, LocaleListEntry).ID)
        End Function
    End Class

    Private Sub GenerateCulturesTable()
        mCultureList.Clear()
        Dim id = 0

        mCultureList.Add(New LocaleListEntry With {
                .ID = id,
                .Text = My.Resources.LocaleConfigForm_UsingWindowsLocaleSettings,
                .Value = Options.Instance.SystemLocale,
                .Active = Options.Instance.SystemLocale = NewLocale
            })

        For Each locale In Internationalisation.Locales.SupportedLocales
            id += 1

            Try
                Dim addCulture = New CultureInfo(locale)
                mCultureList.Add(New LocaleListEntry With {
                    .ID = id,
                    .Text = addCulture.TextInfo.ToTitleCase(addCulture.NativeName),
                    .Value = addCulture.Name,
                    .Active = addCulture.Name = NewLocale
                })
            Catch
            End Try
        Next

        If mPseudoLocalization Then
            Try
                Dim addCulture = New CultureInfo("gsw-LI")
                mCultureList.Add(New LocaleListEntry With {
                    .ID = mCultureList.Count,
                    .Text = "i18n TEST LANGUAGE",
                    .Value = addCulture.Name,
                    .Active = addCulture.Name = NewLocale
                })
            Catch
            End Try
        End If
    End Sub

    Private Sub BtnOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles NextButton.Click
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub BtnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles ExitButton.Click
        DialogResult = DialogResult.Cancel
        mCultureList.FirstOrDefault(Function(x) x.Value = CultureInfo.CurrentUICulture.Name).Active = True
        If Not String.IsNullOrEmpty(NewLocale) Then
            mCultureList.FirstOrDefault(Function(x) x.Value = NewLocale).Active = False
        End If

        Close()
    End Sub

#Region "Drag and Drop"
    Private mMouseDownLocation As Point

    Private Sub BorderPanel_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles BorderPanel.MouseDown
        If e.Button = MouseButtons.Left AndAlso e.Y < 35 Then mMouseDownLocation = e.Location
    End Sub

    Private Sub BorderPanel_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles BorderPanel.MouseMove
        If e.Button <> MouseButtons.Left OrElse e.Y > 35 Then Return
        Dim form = CType(Me, Form)
        form.Left += e.Location.X - mMouseDownLocation.X
        form.Top += e.Location.Y - mMouseDownLocation.Y
    End Sub

    Private Sub BorderPanel_MouseUp(sender As Object, e As MouseEventArgs) Handles BorderPanel.MouseUp
        'following on from a drag and drop of the screen the ListPanel scrolbar become unselectable (mouse scrolling works)
        'but tweaking the vertical scroll value this causes the control to become responsive again, invalidating or refreshing does not work.
        If e.Button = MouseButtons.Left Then
            ListPanel.VerticalScroll.Value = If(ListPanel.VerticalScroll.Value + 1 < ListPanel.VerticalScroll.Maximum, ListPanel.VerticalScroll.Value + 1, ListPanel.VerticalScroll.Maximum)
            ListPanel.VerticalScroll.Value = If(ListPanel.VerticalScroll.Value - 1 > ListPanel.VerticalScroll.Minimum, ListPanel.VerticalScroll.Value - 1, ListPanel.VerticalScroll.Minimum)
        End If
    End Sub

    Private Sub SelectLanguageForm_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
        Select Case e.KeyCode
            Case Keys.Down
                SelectNextLocale(1)
            Case Keys.Up
                SelectNextLocale(-1)
        End Select
    End Sub

    Private Sub SelectNextLocale(dir As Integer)

        Dim oldCulture = mCultureList.FirstOrDefault(Function(x) x.Active)
        Dim index = oldCulture.ID + dir
        Dim label = TryCast(ListPanel.Controls.Find($"L{index}", True).FirstOrDefault, Label)
        If label Is Nothing Then Return
        ListPanel.ScrollControlIntoView(label)
        L_Click(label, EventArgs.Empty)

    End Sub

#End Region
End Class
